ItemPool.py

```diff
@@ -813,524 +323,253 @@ def collect_heart_container(world, pool):
+
+        # Adult Trade Item
+        elif location.vanilla_item == 'Pocket Egg':
+            earliest_trade = trade_item_options.index(world.settings.logic_earliest_adult_trade)
+            latest_trade = trade_item_options.index(world.settings.logic_latest_adult_trade)
+            if earliest_trade > latest_trade:
+                earliest_trade, latest_trade = latest_trade, earliest_trade
+            item = random.choice(trade_items[earliest_trade:latest_trade + 1])
+            world.selected_adult_trade_item = item
+            shuffle_item = True
+
+        # Thieves' Hideout
+        elif location.vanilla_item == 'Small Key (Thieves Hideout)':
+            shuffle_item = world.settings.shuffle_hideoutkeys in ['any_dungeon', 'overworld', 'keysanity']
+            if (world.settings.gerudo_fortress == 'open'
+                    or world.settings.gerudo_fortress == 'fast' and location.name != 'Hideout Jail Guard (1 Torch)'):
+                item = 'Recovery Heart'
+                shuffle_item = False
+            if shuffle_item and world.settings.gerudo_fortress == 'normal' and 'Thieves Hideout' in world.settings.key_rings:
+                item = get_junk_item()[0] if 'Small Key Ring (Thieves Hideout)' in pool else 'Small Key Ring (Thieves Hideout)'
+
+        # Dungeon Items
+        elif location.dungeon is not None:
+            dungeon = location.dungeon
+            shuffle_setting = None
+            dungeon_collection = None
+
+            # Boss Key
+            if location.vanilla_item == dungeon.item_name("Boss Key"):
+                shuffle_setting = world.settings.shuffle_bosskeys if dungeon.name != 'Ganons Castle' else world.settings.shuffle_ganon_bosskey
+                dungeon_collection = dungeon.boss_key
+                if shuffle_setting == 'vanilla':
+                    shuffle_item = False
+            # Map or Compass
+            elif location.vanilla_item in [dungeon.item_name("Map"), dungeon.item_name("Compass")]:
+                shuffle_setting = world.settings.shuffle_mapcompass
+                dungeon_collection = dungeon.dungeon_items
+                if shuffle_setting == 'vanilla':
+                    shuffle_item = False
+            # Small Key
+            elif location.vanilla_item == dungeon.item_name("Small Key"):
+                shuffle_setting = world.settings.shuffle_smallkeys
+                dungeon_collection = dungeon.small_keys
+                if shuffle_setting == 'vanilla':
+                    shuffle_item = False
+                elif dungeon.name in world.settings.key_rings and not dungeon.small_keys:
+                    item = dungeon.item_name("Small Key Ring")
+                elif dungeon.name in world.settings.key_rings:
+                    item = get_junk_item()[0]
+                    shuffle_item = True
+            # Any other item in a dungeon.
+            elif location.type in ["Chest", "NPC", "Song", "Collectable", "Cutscene", "BossHeart"]:
```

Comment:
"Scrub" should be added to this list, right?

Reply:
Scrubs are handled earlier.